numpy/lib/function_base.py

```diff
@@ -3541,7 +3541,7 @@ def _percentile(a, q, axis=None, out=None,
     elif interpolation == 'higher':
         indices = ceil(indices).astype(intp)
     elif interpolation == 'midpoint':
-        indices = floor(indices) + 0.5
+        indices = 0.5 * (floor(indices) + ceil(indices))
```

Comment:
If you look at the docs, `midpoint` is just half-way between the surrounding points, making it effectively a special case of 'linear'. The previous definition of `floor(indices) + 0.5` worked fine unless `indices` was an integer. Float indices are used to do a linear interpolation on the bounding elements. My initial gut response was to do

    f = floor(indices)
    if f == indices:
        indices = indices.astype(intp)
    else:
        indices = f

The problem is of course that some of the indices may be integers and some may not. In this case, we have to do the linear interpolation on all of them, which I have set up correctly. Basically, if the index has a factional part, take the midpoint of the bounding element, otherwise interpolate, but with both bounds being the same.